### PR TITLE
Build out a backend schema editor

### PR DESCRIPTION
**Pull request by jeancochrane**
## Overview

This PR pulls out the Ashlar-specific components of the DRIVER schema editor and incorporates them as a backend to the Ashlar reference implementation.

### Checklist

- [X] PR has a descriptive enough title to be useful in changelogs

### Demo

Navigate to `localhost:9000`, log in, and view the editor up and running:

ashlar-editor

### Notes

- The diff here is regrettably large, since I'm essentially porting another app into this repo. You can safely ignore changes in the following repos, since they're almost exclusively subtractive from what exists in DRIVER:
    - `ashlar/ashlar_auth` (authentication component of Ashlar, forked from `driver/app/driver_auth`)
    - `schema_editor`
- It strikes me that the editor should probably live in its own repo in the same way that Ashlar lives in its own repo, separate from this implementation. Unfortunately there's some hard-wiring between both services that is leftover from DRIVER; once the first draft of the reference implementation is reasonably complete, it'd be nice to take a step back and fully abstract these components out into their own repos.

## Testing Instructions

Run the tests:

- Update builds and dependencies with `./scripts/update` 
- Run `./scripts/test` to run all tests

Run the development server and poke around to make sure everything looks good:

- Update builds and dependencies with `./scripts/update`
- Run the development server with `./scripts/server`
- Navigate to the schema editor at the origin `localhost:9000`, login with the username `admin` and the password `admin`, and browse around

Closes https://github.com/azavea/ashlar-2018-fellowship/issues/5.

### CODE REVIEW

File: ashlar/ashlar_auth/migrations/0001_initial.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    admin_user = User.objects.filter(username=ADMIN_USER['username'])
+    admin_group = Group.objects.get(name=settings.USER_GROUPS['ADMIN'])
+    if len(admin_user) == 0:
+        default_admin = User(**ADMIN_USER)
+        default_admin.set_password(settings.DEFAULT_ADMIN_PASSWORD)
+        default_admin.save()
+        default_admin.groups.add(admin_group)
+        default_admin.save()
+
+def delete_default_admin(apps, schema_editor):
```
Comment (ddohler) [inline ashlar/ashlar_auth/migrations/0001_initial.py:46]: Nice job adding reverse migrations here; that's always a good idea.

File: ashlar/ashlar_auth/migrations/0002_grant_existing_users_permissions.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    try:
+        public_group = Group.objects.get(name=settings.USER_GROUPS['READ_ONLY'])
+        non_admin_users = User.objects.exclude(username=settings.DEFAULT_ADMIN_USERNAME)
+        for user in non_admin_users:
+            user.groups.add(public_group)
+            user.save()
+    except (User.DoesNotExist, Group.DoesNotExist):
+        pass
+
+def drop_default_permissions(apps, schema_editor):
```
Comment (ddohler) [inline ashlar/ashlar_auth/migrations/0002_grant_existing_users_permissions.py]: You can use `RunPython.noop()` here instead. https://docs.djangoproject.com/en/2.0/ref/migration-operations/#django.db.migrations.operations.RunPython.noop

File: ashlar/ashlar_server/management/commands/_wait_for_db.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+import time
```
Comment (ddohler) [inline ashlar/ashlar_server/management/commands/_wait_for_db.py]: This is exactly the right idea for solving this issue, but I think you might find it more effective to move it one step higher up the chain of command. That is, if you move this functionality into `./scripts/server`, then you'll only have to write the logic once, whereas with this approach you're in a situation where you'll have to add a little wrapper command around every single management command we might want to run autonomously.

The downside is that the `./scripts/server` would no longer be a simple `docker-compose up` but would instead have to bring up the `db` service, then try to run `./manage.py dbshell` (or runserver) until it succeeds, and only then launch the remaining services.

File: schema_editor/app/.buildignore [hunk 4]
```diff
@@ -0,0 +1 @@
+*.coffee
```
Comment (ddohler) [inline schema_editor/app/.buildignore]: Does this need to be here? We don't use Coffeescript, to my knowledge.

File: schema_editor/app/scripts/resources/blackspotconfig-service.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+(function () {
```
Comment (ddohler) [inline schema_editor/app/scripts/resources/blackspotconfig-service.js]: This is DRIVER-specific and can be removed.

File: schema_editor/app/scripts/resources/recordcosts-service.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+(function() {
```
Comment (ddohler) [inline schema_editor/app/scripts/resources/recordcosts-service.js]: This is DRIVER-specific and can be removed.

File: schema_editor/app/scripts/views/recordtype/related-aggregates-directive.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
+(function() {
```
Comment (ddohler) [inline schema_editor/app/scripts/views/recordtype/related-aggregates-directive.js]: I'm fairly certain that this is DRIVER-specific and can be removed.

File: schema_editor/app/scripts/views/recordtype/related-aggregates-partial.html [hunk 8]
```diff
@@ -0,0 +1,73 @@
+<div class="form-area">
```
Comment (ddohler) [inline schema_editor/app/scripts/views/recordtype/related-aggregates-partial.html]: Same comment as previous.

File: schema_editor/prototype/add_new_related_content_type.html [hunk 9]
```diff
@@ -0,0 +1,111 @@
+<!DOCTYPE html>
```
Comment (ddohler) [inline schema_editor/prototype/add_new_related_content_type.html]: Is this code related to the first Flask-based version? If so, I'm wondering whether it needs to stick around in the repo. Perhaps it would be better to have it on its own branch?

**Review by ddohler — CHANGES_REQUESTED** [review]
Great start! I had some trouble getting this to work; I'm seeing 404 requests for `config.js`, and that is (I think) causing the rest of the app to fail to initialize with `Module 'ase.config' is not available!`. Aside from that, most of my comments had to do with files that are probably DRIVER-specific enough that they can be removed.

**Comment by jeancochrane** [comment]
Great catch on the config file, @ddohler! Looks like there was a `.gitignore` file in the schema editor app directory that I unknowingly copied over. I added the right `config.js` file to version control, so you should be able to run the app now. (We'll have to remove the config from version control and provision it differently when deploying, but for local development I figured it was fine to leave in there for now -- let me know if you disagree.)

File: schema_editor/prototype/add_new_related_content_type.html (hunk 9 above)
Comment (jeancochrane) [reply schema_editor/prototype/add_new_related_content_type.html]: I copied the `prototype` directory over [from DRIVER](https://github.com/WorldBank-Transport/DRIVER/tree/a33fdd503972dc0f5c87d52f3844d5a19d4dc89a/schema_editor/prototype) when I first got working on this branch, but after spending a lot of time with the schema editor app I don't believe `prototype` does anything. I'll go ahead and remove it.

File: schema_editor/app/.buildignore (hunk 4 above)
Comment (jeancochrane) [reply schema_editor/app/.buildignore]: Good catch! These dotfiles from DRIVER are easy to miss.

File: ashlar/ashlar_auth/migrations/0001_initial.py (hunk 1 above)
Comment (jeancochrane) [reply ashlar/ashlar_auth/migrations/0001_initial.py:46]: I wish I could take credit for it! @flibbertigibbet wrote these migrations for DRIVER: https://github.com/WorldBank-Transport/DRIVER/blob/a33fdd503972dc0f5c87d52f3844d5a19d4dc89a/app/driver_auth/migrations/0001_initial.py#L44-L49 :wink: I think at some point this authentication logic should move into Ashlar core, however -- see https://github.com/azavea/ashlar-2018-fellowship/issues/22.

File: ashlar/ashlar_server/management/commands/_wait_for_db.py (hunk 3 above)
Comment (jeancochrane) [reply ashlar/ashlar_server/management/commands/_wait_for_db.py]: Yep, `./scripts/server` is definitely the right level of abstraction for the database polling logic. Good thinking!

**Comment by ddohler** [comment]
The `config.js` fixes helped--I was able to view the login page. But now when I try to log in, it sends an `OPTIONS` request but then no follow-up request, and then there's a console error saying
```
Error: data is null
AuthService/module.authenticate/<@http://localhost:9000/scripts/auth/auth-service.js:98:1
```

I'm guessing from the `OPTIONS` request that this is a CORS problem?

**Comment by jeancochrane** [comment]
That's strange -- we should be whitelisting anything coming from localhost:

https://github.com/azavea/ashlar-blueprint/blob/96a7bf339093b151059f11115e6526a1abc7b081/ashlar/ashlar_server/settings.py#L117-L120

Is there any more output from the console? It looks like the `auth-service.js` script is complaining, I wonder if we can get a full traceback there.

**Comment by ddohler** [comment]
I tried accessing it both ways, using my internal Azavea IP (`lima.internal.azavea.com`) and then also by port-forwarding to my laptop and accessing `localhost:9000`. There is a longer stack trace but I cut it out because it seemed to be just a bunch of Angular promise resolution stuff.

File: scripts/server [hunk 10]
```diff
@@ -6,29 +6,115 @@ if [[ -n "${ASHLAR_BLUEPRINT_DEBUG}" ]]; then
...
+function run_ashlar_service() {
+    #
+    # Run the Ashlar server instance.
+    #
+    checkport 8000
+    init_db
+    if [ $# -eq 0 ];
+    then
+        # No arguments -- default to running a Django dev server
+        docker-compose run --rm --service-ports ashlar manage.py runserver 0.0.0.0:8000
```
Comment (ddohler) [inline scripts/server]: Since `manage.py runserver 0.0.0.0:8000` is the `command:` for the Ashlar service, I think you should just be able to do `docker-compose up ashlar` here. (Same comment for the editor down on 68).

File: scripts/server [hunk 11]
```diff
@@ -6,29 +6,115 @@ if [[ -n "${ASHLAR_BLUEPRINT_DEBUG}" ]]; then
...
+    fi
+}
+
+function run_editor_service() {
+    #
+    # Run the schema editor app.
+    #
+    checkport 8000
+    init_db
+    docker-compose up -d ashlar
```
Comment (ddohler) [inline scripts/server:67]: This and `init_db` tend to leave running containers lying around, i.e. if I run `./scripts/server` and then `Ctrl-C` it, and then try to run it again, I'll get an error message about the port being in use until I run `docker ps` to find the running containers and kill them. I think it would be a good idea to trap termination signals and then kill (and `rm`) the db and ashlar containers before exiting. http://tldp.org/LDP/Bash-Beginners-Guide/html/sect_12_02.html

File: ashlar/ashlar_server/settings.py [hunk 12]
```diff
@@ -110,5 +114,34 @@
 
 STATIC_URL = '/static/'
 
+# Allow CORS requests from any port on the loopback interface
+CORS_ORIGIN_REGEX_WHITELIST = (
+    r'^http://127.0.0.1:\d+',
```
Comment (ddohler) [inline ashlar/ashlar_server/settings.py]: As we discovered, this needs to allow (at the least) `localhost` as well, but it may be worth finding a way to disable it completely on develop (perhaps keying off `DEBUG`?).

File: schema_editor/app/index.html [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+    <script src="bower_components/angular-ui-router/release/angular-ui-router.js"></script>
+    <script src="bower_components/json-editor/dist/jsoneditor.js"></script>
+    <script src="bower_components/angular-local-storage/dist/angular-local-storage.js"></script>
+    <script src="bower_components/angular-bootstrap/ui-bootstrap-tpls.js"></script>
+    <script src="bower_components/lodash/lodash.js"></script>
+    <!-- endbower -->
+    <!-- endbuild -->
+
+        <!-- build:js({.tmp,app}) scripts/scripts.js -->
+        <script src="scripts/config.js"></script>
```
Comment (ddohler) [inline schema_editor/app/index.html:47]: This is still generating a few 404s from things that were removed in other places but are still listed here.

File: scripts/server (hunk 11 above)
Comment (ddohler) [reply scripts/server:67]: Also, as we discussed, it might be good to try out the Azavea PostGIS image to see if it'll be faster to boot up and also improve the data persistence issues we were seeing. It might be possible to switch back to vanilla `docker-compose up` and strip out some of this DB-initialization logic entirely if the image works better.

**Review by ddohler — CHANGES_REQUESTED** [review]
I think this is really close; there are a few more areas that need improvement that we talked about earlier today, but I think this should be good to go after those are dealt with.

File: scripts/server (hunk 11 above)
Comment (jeancochrane) [reply scripts/server:67]: The Azavea PostGIS image loads much faster! I'm going to keep the database-initialization logic for now, since initial testing indicates that it can take between 1-3 seconds longer to load the DB than the app on my machine. But still, that's nearly 10 seconds faster, which is going to be much less frustrating for development.

File: schema_editor/app/index.html (hunk 13 above)
Comment (jeancochrane) [reply schema_editor/app/index.html:47]: Gotcha -- removed the three scripts that were missing.

**Comment by jeancochrane** [comment]
Addressed your latest round of comments @ddohler! Let me know if there's anything else that jumps out to you.

**Review by ddohler — APPROVED** [review]
I think this is good to go! Thanks for working through all the issues!

**Comment by jeancochrane** [comment]
Fantastic, thank you! Merging away.
